flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala

```diff
@@ -411,6 +411,23 @@ class TaskManager(
             log.debug(s"Cannot find task to fail for execution ${executionID})")
           }
 
+        // stops a task
+        case StopTask(executionID) =>
+          val task = runningTasks.get(executionID)
+          if (task != null) {
+            try {
+              task.stopExecution()
```

Comment:
About Kafka, I think the approach would be to have `cancel()` signal to the source that it should close itself, then have the actual shutdown/cleanup in `close()` of RichFunction. What do you think about this @rmetzger ?